rewrite-java/src/main/java/org/openrewrite/java/RenameVariable.java

```diff
@@ -38,35 +45,95 @@ public RenameVariable(J.VariableDeclarations.NamedVariable variable, String toNa
 
     private class RenameVariableByCursor extends JavaIsoVisitor<P> {
         private final Cursor scope;
-        private boolean isConstant = false;
+        private final Stack<Tree> currentNameScope;
+        private @Nullable Tree currentBlock;
 
         public RenameVariableByCursor(Cursor scope) {
             this.scope = scope;
+            this.currentNameScope = new Stack<>();
+        }
+
+        @Override
+        public J.Block visitBlock(J.Block block, P p) {
+            currentBlock = block;
+            return super.visitBlock(block, p);
         }
 
         @Override
-        public J.CompilationUnit visitCompilationUnit(J.CompilationUnit cu, P p) {
-            return super.visitCompilationUnit(cu, p);
+        public J.ClassDeclaration visitClassDeclaration(J.ClassDeclaration classDecl, P p) {
+            // Variables declared in a class declaration are scoped to the class regardless of statement order.
+            List<Statement> declaringStatement = classDecl.getBody().getStatements().stream()
```

Comment:
Avoid uses of `stream.collect` in recipes in hot paths. This is an expensive operation.

Reply:
I'll avoid .collect in the future. I realized I don't need the list any more, so I replaced the .collect with a boolean.